pallene/assignment_conversion.lua

```diff
@@ -143,27 +148,69 @@ function Converter:apply_transformations()
                 { value = decl._type }
             )
             self:add_box_type(decl.loc, typ)
-            decl._type = typ
 
             local init_exp_update = self.update_init_exp_of_decl[decl]
 
             if init_exp_update then
                 local old_exp = init_exp_update.node
+                decl._type = typ
                 local update  = init_exp_update.update_fn
 
                 local new_node = ast.Exp.InitList(old_exp.loc, {{ name = "value", exp = old_exp }})
                 new_node._type = typ
                 update(new_node)
+
+            elseif self.lambda_of_param[decl] then
+                --- Function parameters that are captured are implementing by "proxy"-ing them.
+                --- Consider the following function that returns a closure:
+                --- ```
+                --- function m.foo(n: integer)
+                ---   -- capture n
+                --- end
+                --- ```
+                --- Since we cannot transform the declaration node of a function parameter,
+                --- we create a variable to represent the boxed parameter which can be captured.
+                --- ```
+                --- function m.foo(n: integer)
+                ---   local $n: $T = { value = n }
+                ---   -- capture and mutate $n
+                --- end
+                local param = ast.Exp.Var(decl.loc, ast.Var.Name(decl.loc, decl.name))
+                param.var._def = checker.Def.Variable(decl)
+                param.var._type = decl._type
+                param._type = decl._type
+                local decl_lhs = ast.Decl.Decl(decl.loc, "$"..decl.name, typ)
+                local decl_rhs = ast.Exp.InitList(decl.loc, {{ name = "value", exp = param }})
+                decl_rhs._type = typ
+                decl_lhs._type = typ
+
+                local new_node = ast.Stat.Decl(decl.loc, { decl_lhs }, { decl_rhs })
+
+                local lambda = self.lambda_of_param[decl]
+                table.insert(lambda.body.stats, 1, new_node)
+                proxy_decl_of_param[decl] = decl_lhs
+
             else
                 error("upvalues that are not initialized upon declaration cannot be captured.")
             end
 
             for _, node_update in ipairs(self.update_ref_of_decl[decl]) do
                 local old_var = node_update.node
+                local loc     = old_var.loc
                 local update  = node_update.update_fn
 
-                local loc = old_var.loc
-                local new_node = ast.Var.Dot(loc, ast.Exp.Var(loc, old_var), "value")
+                local dot_exp
+                if proxy_decl_of_param[decl] then
+                    -- references to captured parameters get replaced by references to `value` field of
+                    -- their proxy variables.
+                    local proxy_var = ast.Var.Name(old_var.loc, "$"..decl.name)
+                    proxy_var._def = checker.Def.Variable(proxy_decl_of_param[decl])
```

Comment:
Technically this proxy_var and this def are the same for all the places that use the captured parameter. Should we reuse them instead of recreating them every time?

Reply:
I decided to keep it that way because in a real AST those nodes would be repeated, right?
But I do agree that in this case we can use the same node. I'll change that 👍

Reply:
I wonder if we could replace the `proxy_decl_of_param` by a `proxy_var_of_param`, or something of the sort.

Reply:
Another reason I think using the same node might not be the best is because of the `loc` field. I think it's best to keep the `loc`s as close as possible to the original var nodes.